SignalServiceKit/src/Storage/FullTextSearchFinder.swift

```diff
@@ -0,0 +1,141 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+// Create a searchable index for objects of type T
+public class SearchIndexer<T> {
+
+    private let indexBlock: (T) -> String
+
+    public init(indexBlock: @escaping (T) -> String) {
+        self.indexBlock = indexBlock
+    }
+
+    public func index(_ item: T) -> String {
+        return indexBlock(item)
+    }
+}
+
+@objc
+public class FullTextSearchFinder: NSObject {
+
+    public func enumerateObjects(searchText: String, transaction: YapDatabaseReadTransaction, block: @escaping (Any, String) -> Void) {
+        guard let ext: YapDatabaseFullTextSearchTransaction = ext(transaction: transaction) else {
+            assertionFailure("ext was unexpectedly nil")
+            return
+        }
+
+        let normalized = FullTextSearchFinder.normalize(text: searchText)
+
+        // We want a forgiving query for phone numbers
+        // TODO a stricter "whole word" query for body text?
+        let prefixQuery = "*\(normalized)*"
+
+        // (snippet: String, collection: String, key: String, object: Any, stop: UnsafeMutablePointer<ObjCBool>)
+        ext.enumerateKeysAndObjects(matching: prefixQuery, with: nil) { (snippet: String, _: String, _: String, object: Any, _: UnsafeMutablePointer<ObjCBool>) in
+            block(object, snippet)
+        }
+    }
+
+    private func ext(transaction: YapDatabaseReadTransaction) -> YapDatabaseFullTextSearchTransaction? {
+        return transaction.ext(FullTextSearchFinder.dbExtensionName) as? YapDatabaseFullTextSearchTransaction
+    }
+
+    // Mark: Index Building
+
+    private class var contactsManager: ContactsManagerProtocol {
+        return TextSecureKitEnv.shared().contactsManager
+    }
+
+    private class func normalize(text: String) -> String {
+        var normalized: String = text.trimmingCharacters(in: .whitespacesAndNewlines)
+
+        // Remove any phone number formatting from the search terms
+        let nonformattingScalars = normalized.unicodeScalars.lazy.filter {
+            !CharacterSet.punctuationCharacters.contains($0)
+        }
+
+        normalized = String(String.UnicodeScalarView(nonformattingScalars))
+
+        return normalized
+    }
+
+    private static let groupThreadIndexer: SearchIndexer<TSGroupThread> = SearchIndexer { (groupThread: TSGroupThread) in
+        let groupName = groupThread.groupModel.groupName ?? ""
+
+        let memberStrings = groupThread.groupModel.groupMemberIds.map { recipientId in
+            recipientIndexer.index(recipientId)
+        }.joined(separator: " ")
+
+        let searchableContent = "\(groupName) \(memberStrings)"
```

Comment:
So with this arrangement, could a snippet span a group name and some other content?  Should we use some kind of delimiting token and chop up snippets using those delimiters?

Reply:
If I have a group called "Book Club" and a member named "Alice"
When I search "Book Alice"

I (personally) would expect "Book Club" to show up, which is the current behavior. It sounds like you're suggesting the opposite.

Can you give me an example in terms of user behavior and the undesirable outcome the existing approach could produce?